src/mlpack/methods/ann/gan_impl.hpp

```diff
@@ -0,0 +1,332 @@
+        disBatchSize - 1)));
+
+    responses.cols(0, disBatchSize - 1).zeros();
+    responses.cols(disBatchSize, batchSize - 1).ones();
+
+    this->numFunctions = predictors.n_cols;
+    discriminator.numFunctions = predictors.n_cols;
+  }
+  else
+  {
+    predictors.set_size(noiseInSize, batchSize);
+    noiseData.set_size(noiseInSize, batchSize);
+    
+    noiseData = arma::mat(predictors.memptr(), noiseInSize,
+        batchSize, false, false);
+    genFakeData = arma::mat(discriminator.predictors.memptr(), n_rows,
+        batchSize, false, false);
+    GenerateNoise(std::move(genFakeData), std::move(noiseData));
+    
+    responses.ones();
+    this->numFunctions = predictors.n_cols;
+  }
+}
+
+template<typename Generator, typename Discriminator, typename IntializerType>
+template<typename OptimizerType>
+void GenerativeAdversarialNetwork<Generator, Discriminator, IntializerType>
+::Train(OptimizerType& Optimizer)
+{
+  if (!reset)
+    Reset();
+  offset = 0;
+  trainGenerator = false;
+  GenerateData();
+  Optimizer.Optimize(*this, parameter);
+}
+
+template<typename Generator, typename Discriminator, typename IntializerType>
+double GenerativeAdversarialNetwork<Generator, Discriminator, IntializerType>
+::Evaluate(const arma::mat& /*parameters*/,
+    const size_t i, const bool /*deterministic*/)
+{
+  currentInput = this->predictors.unsafe_col(i);
+  currentTarget = this->responses.unsafe_col(i);
+  Forward(std::move(currentInput));
+  double res = discriminator.outputLayer.Forward(std::move(
+      boost::apply_visitor(outputParameterVisitor,
+      discriminator.network.back())), std::move(currentTarget)); 
+  return res;
+}
+
+template<typename Generator, typename Discriminator, typename IntializerType>
+void GenerativeAdversarialNetwork<Generator, Discriminator, IntializerType>
+::Gradient(const arma::mat& parameters, const size_t i, arma::mat& gradient)
+{
+  if (!reset)
+    Reset();
+
+  if (gradient.is_empty())
+  {
+    if (parameter.is_empty())
+    {
+      Reset();
+    }
+    gradient = arma::zeros<arma::mat>(parameter.n_rows, 1);
+  }
+  else
+  {
+    gradient.zeros();
+  }
+
+  // Gradient for generator network
+  gradientGenerator = arma::mat(gradient.memptr(),
+    generator.Parameters().n_rows, 1, false, false);
+
+  // Gradient for discriminator network
+  gradientDiscriminator = arma::mat(gradient.memptr() + gradientGenerator.n_elem,
+    discriminator.Parameters().n_rows, 1, false, false);
+
+  // get the gradients of the discriminator
+  discriminator.Gradient(parameters, i, gradientDiscriminator);
+  if (trainGenerator)
+  {
+    // pass the error from discriminator to generator
+    generator.error = boost::apply_visitor(deltaVisitor,
+        discriminator.network[1]);
+    generator.Backward();
+    generator.ResetGradients(gradientGenerator);
+    // set the current input requrired for gradient computation
+    generator.currentInput = predictors.col(i);
+    generator.Gradient();
```

Comment:
Sorry for the slow response. I don't assert that you did an error since
the sign of `grad[-log(D(G(noise)))]` is equal to the sign of `grad[log(1 - D(G(noise)))]`. I just pointed out that the formula doesn't match. I think we should test both approaches and compare them.
P.S. If you implement something according to another source (i.e. not the GAN papers) you should point that out in the code (i.e. provide the link), otherwise it is hard to verify the code. Moreover, the authors didn't provide any test that verify their approach. Hence, as for me the statement is doubtful and needs verification.